[IPSCCP] Variable not visible at Og:

CarlosAlbertoEnciso opened the pull request:
https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not create a constant expression for the initializer value.

Extend test coverage to include:
- half, bfloat types.
- checks for undef (int32 and ptr).

There is no support for:
- fp128, x86_fp80, ppc_fp128  types.
https://github.com/llvm/llvm-project/issues/88102

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Extend test coverage to include:
- half and bfloat types.
- checks for undef (int32 and ptr).

nikic commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3594,7 +3594,8 @@ DIExpression *llvm::getExpressionForConstant(DIBuilder &DIB, const Constant &C,
   auto *FP = dyn_cast<ConstantFP>(&C);
-  if (FP && (Ty.isFloatTy() || Ty.isDoubleTy())) {
+  if (FP &&
+      (Ty.isFloatTy() || Ty.isDoubleTy() || Ty.isHalfTy() || Ty.isBFloatTy())) {
```
Can you use `isFloatingPointTy()` here to cover *all* FP types?

CarlosAlbertoEnciso replied:
Uploaded a patch to support: `fp128`, `x86_fp80`, `ppc_fp128`.

CarlosAlbertoEnciso replied:
@nikic Very sorry for my late response.
> Edit: Well, looking closer, this is bitcasting the float to an integer.

Looking at the documentation for `bitcastToAPInt`
```
// This function creates an APInt that is just a bit map of the floating
// point constant as it would appear in memory.  It is not a conversion,
// and treating the result as a normal integer is unlikely to be useful.
APInt IEEEFloat::bitcastToAPInt() const {
```
The function does not convert the `float` into an `int`.

SLTozer replied:
The FP value is being bitcast to an int value, so that it can appear in a DWARF expression. To the question about how debug info handles it, my understanding at the moment is that it's up to the consumer. I don't know specifically how different float types of the same size are handled in most debuggers, but there's not much more we can do from this side than provide a bit value and the variable's type AFAICT.

CarlosAlbertoEnciso replied:
@SLTozer Thanks for your additional input on DWARF expression.

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

- Extend test coverage for all float types:
  Half, bfloat, fp128, x86_fp80, ppc_fp128

felipepiovezan reviewed (approved):
this seems reasonable to me, but please wait for a second opinion as well

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Fix assertion detected by the BuildKite pre-commit CI:

llvm::APInt::getZExtValue() const:
`getActiveBits() <= 64 && "Too many bits for uint64_t"' failed.

CarlosAlbertoEnciso commented:
Updated patch to fix assertion detected the BuildKite pre-commit CI:
```
llvm::APInt::getZExtValue() const:
`getActiveBits() <= 64 && "Too many bits for uint64_t"' failed.
```

SLTozer commented on llvm/test/Transforms/SCCP/pr50901.ll:
```diff
@@ -52,13 +52,35 @@
+; CHECK-DAG: ![[DBGB:[0-9]+]] = distinct !DIGlobalVariable(name: "g_ptr_undef"
+; CHECK-DAG: ![[GB:[0-9]+]] = !DIGlobalVariableExpression(var: ![[DBGB]], expr: !DIExpression())
+
+; CHECK: ![[G10:[0-9]+]] = !DIGlobalVariableExpression(var: ![[DBG10:[0-9]+]], expr: !DIExpression(DW_OP_constu, 17293822569102704640, DW_OP_stack_value))
+; CHECK-DAG: ![[DBG10]] = distinct !DIGlobalVariable(name: "g_1010", {{.*}}
+; CHECK: ![[G11:[0-9]+]] = !DIGlobalVariableExpression(var: ![[DBG11:[0-9]+]], expr: !DIExpression(DW_OP_constu, 17293822569102704640, DW_OP_stack_value))
+; CHECK-DAG: ![[DBG11]] = distinct !DIGlobalVariable(name: "g_1111", {{.*}}
+; CHECK: ![[G12:[0-9]+]] = !DIGlobalVariableExpression(var: ![[DBG12:[0-9]+]], expr: !DIExpression(DW_OP_constu, 14480694097861998592, DW_OP_stack_value))
+; CHECK-DAG: ![[DBG12]] = distinct !DIGlobalVariable(name: "g_1212", {{.*}}
```
The values for the >64bit FP values look wrong: the constants for fp128 and ppc_fp128, for example, are equivalent to `0xF000000000000000`, just the upper 64 bits of the original value, and for x86_FP80 the value is `0xC8F5C28F5C28F800`, just the lower 64 bits of the original value (possibly the implementation gives us the last complete 64-bit word?). If there's not a reliable way to make this work for constant values that don't fit in 64 bits, I think it'd be best to drop them.

NB: Off the top of my head DWARF does have representation for values of arbitrary length, but we don't produce them in LLVM until the point where we emit DWARF; either DW_OP_piece (which we exclusively represent with DW_OP_LLVM_fragment until emission) or `DW_OP_implicit_value`, which we emit for DBG_VALUEs with constant values and non-complex expressions. The best choice here would probably be to use `DW_OP_implicit_value`, but I'm not sure how/if we can represent it well in a `!DIExpression` or whether trying to do so could cause errors.

CarlosAlbertoEnciso replied:
Thanks for your detailed comment. You are correct.
In the following code:
```
  if (FP && Ty.isFloatingPointTy()) {
    const APFloat &APF = FP->getValueAPF();
    APInt const &API = APF.bitcastToAPInt();
    if (auto Temp = API.tryZExtValue())
      return DIB.createConstantValueExpression(static_cast<uint64_t>(*Temp));
    else if (auto Temp = API.trySExtValue())
      return DIB.createConstantValueExpression(static_cast<uint64_t>(*Temp));
    return DIB.createConstantValueExpression(*API.getRawData());
```
`tryZExtValue()` and `trySExtValue()` return `std::nullopt` as the `getSignificantBits()` is `>64
`
and then `getRawData()` returns just the upper 64-bits of the original value, which is the value used for the `DW_OP_constu`.

CarlosAlbertoEnciso replied:
I think dropping support for `fp128`, `ppc_fp128` and `x86_fp80` would be a sensible option.
@nikic any thoughts?

nikic replied:
Yeah, I don't think there's value in supporting those. Just dropping the fallback case where tryZExt/trySExt fail should be fine.

Though I also don't think that trySExt makes sense in the context of a float that has been bitcast to an integer. So I think this should be only tryZExt.

CarlosAlbertoEnciso replied:
@nikic Thanks.
I have created a PR to record this limitation: https://github.com/llvm/llvm-project/issues/88102

SLTozer reviewed (commented):
Got an inline comment - the output for >64bit fp values looks broken; I'm fine with just dropping support for those values, though if there's a good fix then that would be ideal.

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Drop support for fp128, ppc_fp128 and x86_fp80 types.

SLTozer commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3586,10 +3586,13 @@ DIExpression *llvm::getExpressionForConstant(DIBuilder &DIB, const Constant &C,
 
   auto *FP = dyn_cast<ConstantFP>(&C);
-  if (FP && (Ty.isFloatTy() || Ty.isDoubleTy())) {
+  if (FP &&
+      (Ty.isBFloatTy() || Ty.isHalfTy() || Ty.isFloatTy() || Ty.isDoubleTy())) {
```
```suggestion
  if (FP && Ty.isFloatingPointTy() && Ty.getScalarSizeInBits() <= 64) {
```
This _could_ be represented with the above instead, which might make the limitation clearer (i.e. size, not FP semantics) to other users - since I don't think it would make any functional change right now though, this is an optional suggestion.

SLTozer commented on llvm/lib/Transforms/Utils/Local.cpp:
```diff
@@ -3586,10 +3586,13 @@ DIExpression *llvm::getExpressionForConstant(DIBuilder &DIB, const Constant &C,
-    return DIB.createConstantValueExpression(
-        APF.bitcastToAPInt().getZExtValue());
+    APInt const &API = APF.bitcastToAPInt();
+    if (auto Temp = API.tryZExtValue())
```
Now that we aren't supporting values >64bits, I think this can revert to using `getZExtValue`, since `tryZExtValue` only returns `std::nullopt` if `getActiveBits() <= 64`.

SLTozer commented on llvm/unittests/Transforms/Utils/LocalTest.cpp:
```diff
@@ -1256,27 +1268,6 @@ TEST(Local, ExpressionForConstant) {
-  Expr = createExpression(ConstantFP::get(BFloatTy, 32), BFloatTy);
-  EXPECT_EQ(Expr, nullptr);
-
-  Type *FP128Ty = Type::getFP128Ty(Context);
-  Expr = createExpression(ConstantFP::get(FP128Ty, 32), FP128Ty);
-  EXPECT_EQ(Expr, nullptr);
-
-  Type *X86_FP80Ty = Type::getX86_FP80Ty(Context);
-  Expr = createExpression(ConstantFP::get(X86_FP80Ty, 32), X86_FP80Ty);
-  EXPECT_EQ(Expr, nullptr);
-
-  Type *PPC_FP128Ty = Type::getPPC_FP128Ty(Context);
-  Expr = createExpression(ConstantFP::get(PPC_FP128Ty, 32), PPC_FP128Ty);
-  EXPECT_EQ(Expr, nullptr);
```
For the >64 bit types, we can reinstate these tests for nullptr expressions.

SLTozer reviewed (approved):
A couple comments inline, other than that LGTM.

CarlosAlbertoEnciso committed:
[IPSCCP] Variable not visible at Og:

https://bugs.llvm.org/show_bug.cgi?id=51559
https://github.com/llvm/llvm-project/issues/50901

IPSCCP pass removes the global variable and does not
create a constant expression for the initializer value.

Address @SLTozer comments:
- Reinstate the test (FP >= 64) for nullptr expressions.
- Change condition to clarify the FP size restrictions.
- Use 'getZExtValue' instead of 'tryZExtValue'.

CarlosAlbertoEnciso commented:
@SLTozer Thanks for your reviews. I will wait for the tests to be completed and then I will commit.
